### PR TITLE
First go at a MOOSE documentation system

### PR DESCRIPTION
**Pull request by aeslaughter**
Alright this is beginning of the MOOSE Documentation system: #6699 
## Initial Implementation
1. Get the basic structure merged. 
2. Setup civet to start building the new site automatically.
3. Start moving/adding content from the current wiki and website, adding features as needed.
## Demo

There is currently a minimal demo available: http://mooseframework.org/docs/markdown_demo/site/

Be sure to look at:
1. http://mooseframework.org/docs/markdown_demo/site/documentation/MooseFlavoredMarkdown/
2. http://mooseframework.org/docs/markdown_demo/site/documentation/systems/framework/Kernels/Overview/
3. http://mooseframework.org/docs/markdown_demo/site/documentation/systems/framework/Kernels/Diffusion

(refs #6699)

### CODE REVIEW

**Comment by friedmud** [comment]
Wow!  So far I like what I see... even looks good on my iPhone.

I'll investigate more when I get home...

**Comment by friedmud** [comment]
Just saw all of the linking you did with Tutorials, Examples and Tests... that is really awesome!

This new documentation system might be a game changer for the project...

File: framework/docs/doxygen/Doxyfile [hunk 1]
```diff
@@ -1838,7 +1838,7 @@ MAN_LINKS              = NO
 # captures the structure of the code including all documentation.
 # The default value is: NO.
 
-GENERATE_XML           = NO
+GENERATE_XML           = YES
```
Comment (jwpeterson) [inline framework/docs/doxygen/Doxyfile]: Probably need to undo this change.

**Comment by jwpeterson** [comment]
FYI @permcody this commit moves the SQA stuff from "doc" to "docs", e.g. `framework/doc/sqa/Makefile -> framework/docs/sqa/Makefile`, not sure if that change needs to be ... documented somewhere else.

File: docs/documentation/generation/MooseFlavoredMarkdown.md [hunk 2]
```diff
@@ -0,0 +1,69 @@
+# Moose Flavored Markdown
+
+## Including MOOSE Source Files
+It is possible to include complete or partial C++ or input files from the local MOOSE repository. The following sections detail the custom
+markdown syntax to needed, including the application of special settings in the form of key, value pairings that are supplied within
+the custom markdown. A complete list of available settings is provided in the [Settings](MooseFlavoredMarkdown.md#optional-settings) the included code.
```
Comment (jwpeterson) [inline docs/documentation/generation/MooseFlavoredMarkdown.md]: typo: settings "of" the included code.

File: docs/documentation/generation/MooseFlavoredMarkdown.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+![](framework/src/kernels/Diffusion.C line=addClassDescription)
+
+### Range Line match
+Code starting and ending on lines containing a string is also possible by using the 'start' and 'end'
+options. If 'start' is omitted then the snippet will start at the beginning of the file. Similarly, if 'end'
+is omiited the snippet will include the remainder of the file content.
+![](test/tests/kernels/simple_diffusion/simple_diffusion.i start=Kernels end=Executioner)
+
+### Class Methods
+By including a method name, in C++ syntax fashion, it is possible to include specific methods from C++ class in MOOSE. For example,
```
Comment (jwpeterson) [inline docs/documentation/generation/MooseFlavoredMarkdown.md]: typo: "from C++ class*es* in MOOSE."

File: docs/documentation/generation/MooseFlavoredMarkdown.md [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+```
+![simple_diffusion.i](test/tests/kernels/simple_diffusion/simple_diffusion.i::Kernels)
+```
+
+![simple_diffusion.i](test/tests/kernels/simple_diffusion/simple_diffusion.i::Kernels repo_link=True)
+
+
+### Optional Settings
+The following options may be passed to control the how the output is formatted.
```
Comment (jwpeterson) [inline docs/documentation/generation/MooseFlavoredMarkdown.md]: typo: "control the how"

File: docs/documentation/generation/Overview.md [hunk 5]
```diff
@@ -0,0 +1,110 @@
+## Summary
+MOOSE includes tools for building autogenerating web-sites for that documents your application. Included
+in this tool is the ability to:
+
+* Add detailed descriptions of custom systems and objects using a [Moose Style Markdown](MooseFlavoredMarkdown.md),
```
Comment (jwpeterson) [inline docs/documentation/generation/Overview.md]: typo: take out "a"

File: docs/documentation/generation/Overview.md [hunk 6]
```diff
@@ -0,0 +1,110 @@
+## Summary
+MOOSE includes tools for building autogenerating web-sites for that documents your application. Included
```
Comment (jwpeterson) [inline docs/documentation/generation/Overview.md]: typo: "web-sites for that documents your application"

File: docs/documentation/generation/Overview.md [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+
+![](framework/src/ics/FunctionIC.C line=addRequiredParam)
+
+### (2) Configuration
+The documentation generation system for MOOSE is built upon [mkdocs](www.mkdocs.org), thus it is configured
+using [YAML](http://yaml.org) files. For MOOSE there are two configuration files to setup: a source and application
+level configuration.
+
+#### Source Configure File
+Inside each source code directory that MOOSE code exists a "config.yml" file must exist. In general, most
```
Comment (jwpeterson) [inline docs/documentation/generation/Overview.md]: Change sentence to "A config.yml file must exist in each directory containing MOOSE source."

File: docs/documentation/generation/Overview.md [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+application will contain a single "config.yml" file and it will exist at the top level. However, if your application
```
Comment (jwpeterson) [inline docs/documentation/generation/Overview.md]: application*s*

File: docs/documentation/generation/Overview.md [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+has a more complex structure (e.g., MOOSE itself has the framework and modules directories) multiple configuration
+files will allow the documenation to be separated---as in MOOSE---into multiple directories.
+
+A typical configuration file will look similar to those of the modules, for example the following is
+the configuation for the Phase Field module.
+
+![](modules/phase_field/config.yml)
+
+The complete list of options and there purpose are as follows.
```
Comment (jwpeterson) [inline docs/documentation/generation/Overview.md]: there -> their

File: docs/documentation/generation/Overview.md [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+### (3) Detailed Descriptions
+It is possible to add a detailed description, using [Moose Flavored Markdown](MooseFlavoredMarkdown.md) by
+creating.
+
+
+## Building Documentation
+
+### Setup
+Currently, the documentation system for MOOSE is under development so the required dependencies are not yet
+part of the repository so some setup is required.
```
Comment (jwpeterson) [inline docs/documentation/generation/Overview.md]: replace "so" -> "and"

File: python/MooseDocs/MooseSubApplicationDocGenerator.py [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+        for key, value in self._syntax.objects().iteritems():
+            src = self._syntax.filenames(key)
+            nodes = yaml_data[key]
+            for node in nodes:
+                if not any([node['name'].startswith(h) for h in hide]):
+                    self._objects.append(MooseObjectInformation(node, src, inputs=inputs, children=children, **self._config))
+
+    def write(self):
+        """
+        Write tye system and object markdown as well as the associated yaml files for mkdocs.
```
Comment (jwpeterson) [inline python/MooseDocs/MooseSubApplicationDocGenerator.py]: Typo "tye"

File: python/MooseDocs/extensions/MooseCppMethod.py [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+import MooseDocs
+from MooseSourcePatternBase import MooseSourcePatternBase
+import utils
+
+class MooseCppMethod(MooseSourcePatternBase):
+    """
+    A markdown extension for including complete source code files.
+    """
+
+    CPP_RE = r'!\[(.*?)\]\((.*\.[Ch])::(\w+)\s*(.*?)\)'
```
Comment (jwpeterson) [inline python/MooseDocs/extensions/MooseCppMethod.py]: Document regex.

File: python/MooseDocs/method.C [hunk 13]
```diff
@@ -0,0 +1,20 @@
+class Foo
```
Comment (jwpeterson) [inline python/MooseDocs/method.C]: This file should be removed.

**Comment by jwpeterson** [comment]
What I like about this:

1. The documentation is truly revision controlled in a simple git repo, not hidden behind a wiki, etc. and backed up by every person who clones MOOSE.
1. The web page it generates looks pretty good.

My concerns are:

1. We're creating a new Python project which will need to be maintained. 
1. We have to use a special patched version of mkdocs, so Andrew's fork will need to be updated/maintained over time.
1. The foo.md files will get out of sync, despite our best efforts, with the classes they document.
1. According to @aeslaughter, it can be somewhat slow to regenerate certain kinds of docs, so we'll need to keep an eye on how much time the build machines are spending do this...
1. How does all this interact with the workshop slides (if at all)?  Are we going to have to regenerate all that stuff... again?

**Comment by aeslaughter** [comment]
> We're creating a new Python project which will need to be maintained.

Well, I hope this will replace the presentation blaster soon (see #5) so it will end up being a conservation of python projects.



> We have to use a special patched version of mkdocs, so Andrew's fork will need to be updated/maintained over time.

They are proposing an API for extending there project, depending how important this issue is I can work on pushing that API along to meet our needs and theirs.

https://github.com/mkdocs/mkdocs/wiki/mkdocs-Plugin-API/_edit



> The foo.md files will get out of sync, despite our best efforts, with the classes they document.

What aspect will get out of sync? All of the parameter tables and links are generated from the source code. As reviewers we will need to make sure that when an object is changed that the person submitting the change has also updated the detailed description. Eventually, anyone adding a new object will have a failed PR if they do not include an md file explaining there code. This will be more work for the reviewers but create a traceable path forward for requiring documentation, which is far better than we have now (i.e., no requirement).

> According to @aeslaughter, it can be somewhat slow to regenerate certain kinds of docs, so we'll need to keep an eye on how much time the build machines are spending do this...

The only bit that is slow is extracting methods using clang (a few seconds), I have provided other quicker text based methods that should be used unless you really need to use the clang. 

> How does all this interact with the workshop slides (if at all)? Are we going to have to regenerate all that stuff... again?

Eventually (I have a prototype) this will automatically generate slides for every page, so we can then just create a page for the training that contains a complete outline for the training and the slides would be generated automatically. We will have a concrete place for people to go and look at slides, no more python script for people to run.

**Comment by jwpeterson** [comment]
> We will have a concrete place for people to go and look at slides, no more python script for people to run.

Major :+1: for this, especially if it could also generate a PDF.

**Comment by aeslaughter** [comment]
> Major 👍 for this, especially if it could also generate a PDF.

Also on the list of things to add.

**Comment by friedmud** [comment]
I like the conversation here...

Let me add an issue: I kind of don't want the presentation slides to be
automatically generated from this documentation.  We've kind of relied a
little to heavily on that and it's chained us down to the idea that all of
documentation needs to be short-form bulleted lists with pictures.

I'm really hoping that with this new documentation system I can write some
_prose_.  I want to write out long-form descriptions of systems and classes
and how they fit into the overall design goals of MOOSE.  I want to write
down all of the things we normally say during a training class.  I actually
want the documentation to be about "why" as well as "how".  I feel like we
miss out on a lot of users because they don't "get" MOOSE...

I think our tutorials should be paired down a lot and focused a bit more on
the task at hand.  Yes, this will create more of a maintenance burden
(having two similar looking pieces of documentation at times) but MOOSE is
stabilizing and things are changing much more slowly now so I feel like it
won't pull us down.

BTW: I also agree with the requirement to document things before they are
merged.  We're going to have to stay on top of this in the future.

Just some $0.02

**Comment by aeslaughter** [comment]
With this system we control the markdown syntax completely, so we could
build some sort of tagging system that is placed in the markdown that
communicates what part of the document is put into slides.

I also don't mind not building the slides for every page, but just for the
tutorial(s) and training information, which may or may not be the same
markdown.

**Comment by friedmud** [comment]
@aeslaughter What is your near-term plan for this?  Should we put off merging it until the tiger team so we can talk more about the implications?

I'm also wary of merging this with @permcody out.

To be clear: I have no issue with what's here... the question is more about the design and long-term consequences.  We need to really be sure that this is what we want to do before going forward...

**Comment by aeslaughter** [comment]
@friedmud 
My plan was to:

1. Merge this
2. Get Civet building and copying the docs to mooseframework.org/demo_or_something
3. Document Pika using the system to figure out what we need in Stork
4. Start documenting code and add features as needed.

**Comment by dschwen** [comment]
Anything happening here? What are we waiting for?

**Comment by friedmud** [comment]
I was either waiting for @permcody or for the Tiger Team.  This is a BIG decision and I want to make sure this is the way we want to go.

We're going to put hundreds / thousands of man-hours into putting information into this system... we need to get it right.

**Comment by permcody** [comment]
I'm behind this, but I haven't looked in depth at the code or anything. I just watched @aeslaughter's 20 min presentation is all. He's still on vacation though so we'll have to at least wait until he returns.

**Comment by aeslaughter** [comment]
We should have a meeting about this on Monday. I would like to see this get merged soon if possible so I can start using it to see what needs to be improved.

However, I also don't want people to start going too crazy with it yet until we get the design nailed down, which may include me adding some  code to mkdocs itself, I have been chatting with there developers quite a bit. So, that bit might need to come before this is merged, it depends on how much more time I should be spending on this and who is paying for it.

**Comment by friedmud** [comment]
I'm definitely available on Monday.  Just let me know when.

Derek

**Comment by aeslaughter** [comment]
I am making this work better with out-of-the-box mkdocs, just to avoid the custom package as I continue to work with the mkdocs team.

**Comment by aeslaughter** [comment]
see #7380
